### ISSUE TITLE
Trouble reproducing Fig. 2 iPSC interpolation (day 0/3 → day 1/2)

### ISSUE THREAD

**Issue by jhxu003** [siyuh/Squidiff#8]
Hi, we’re trying to reproduce the iPSC experiment from Fig. 2—training only on day 0 and day 3 cells, then interpolating day 1 and day 2—but we’re not getting the reported behavior.

Because no checkpoint for this setting was provided, we trained our own model from scratch. Below is exactly what we did:

### Data processing

1. Start from the iPSC dataset referenced in the paper (Zenodo record 3625024).
2. Filter cells to keep only day 0–day 3 samples, remove low-quality cells (total counts ≥ 1,000; 200 ≤ genes expressed ≤ 8,000; mitochondrial fraction ≤ 20%).
3. Normalize per cell to 1 × 10⁴ counts, log1p transform.
4. Select 203 highly variable genes (as described in the paper) and force-include the marker genes mentioned around Fig. 2 (NANOG, T, GATA6, POU5F1, HAND1, MKI67, ETV2, CLDN5, PECAM1, SOX17, COL4A1, DCN, ACTA2, PDGFRB, FOXA2, EOMES, MIXL1, COL5A1, CXCR4, NR2F2, GJA4, PGF, RGS16, ADAMTS18).
5. Training data: random balanced sample of 1,200 cells from day 0 and 1,200 cells from day 3 (so only day 0/3 go into training).
6. Evaluation data: separate balanced sample of 600 cells per day across day 0–3 (used purely for interpolation/evaluation).

### Training configuration

CUDA_VISIBLE_DEVICES=2 python train_squidiff.py \
       --logger_path logs/ipsc_fig2_run \
       --resume_checkpoint checkpoints/ipsc_fig2_run \
       --data_path datasets/ipsc_fig2_train.h5ad \
       --gene_size 203 \
       --output_dim 203 \
       --num_layers 3 \
       --batch_size 64 \
       --microbatch -1 \
       --lr 1e-4 \
       --diffusion_steps 1000 \
       --lr_anneal_steps 100000 \
       --log_interval 200 \
       --save_interval 2000 \
       --use_drug_structure False

Training with loss ≈ 0.03, and gradients looked stable.

### Evaluation procedure

- Encode the evaluation split (day 0–3), treat Group 0 as “start” (day 0) and Group 3 as “target” (day 3).
- Compute Δz in the semantic latent space and interpolate at fractions corresponding to days 1 and 2.
- Generate 1,000 cells per interpolated day and compare to the held-out real day 1/2 cells:
      - UMAP overlay (real vs generated)
      - Mean-expression scatter
      - Flattened Pearson/R²/MAE/RMSE.

### Observed results

For both day 1 and day 2 comparisons we get:

- Mean-level Pearson ≈ 0.17, R² ≈ –0.04
- Flattened R² ≈ –0.06, RMSE ≈ 0.58, MAE ≈ 0.30
- UMAPs show almost no overlap between real and generated cells; the model behaves like a noisy baseline rather than capturing the day 1/2 state.

### Paper reference

Fig. 2b–c and Supplementary Fig. 2 report that Squidiff, trained only on days 0 and 3, “accurately predicts single-cell transcriptomics from day 1 to day 3,” showing smooth differentiation trajectories. We aren’t seeing that behavior—the interpolated cells stay far from the real day 1/2 distribution.

### Questions

1. Are we missing any preprocessing steps (e.g., scaling, gene ordering, semantic-encoder options) that were crucial in your Fig. 2 notebook?
2. Do the iPSC experiments use different architecture settings (gene_size/output_dim, num_layers, channels, EMA, etc.) or training tricks (e.g., curriculum, lower LR) that we should replicate?
3. Is the new Figshare checkpoint (“VO_diff_model.pt”) intended for this iPSC interpolation, or is there a separate checkpoint for Fig. 2 that we should request?
4. Any guidance on diagnosing why our R² stays negative despite following the published description?

Thanks in advance for pointing out what we’re missing—we’d love to line up our replication with the paper’s results.

**Comment by baldmk** [comment]
Hello! That's a great question!
I noticed that neither your preprocessing pipeline nor the one described in the paper mentions z-score scaling. Therefore, before reproducing Figure 2, I wondered if this step was actually required but overlooked and unmentioned in the paper—and since you didn’t perform it either, it might have led to a significant deviation in the evaluation scores.
Initially, I hypothesized that applying scaling would be necessary to achieve the results reported in the paper, and the suboptimal performance you encountered was due to omitting this step. To investigate the root cause, I prepared two sets of datasets during preprocessing:

1. One set with z-score scaling applied to both the training and test datasets;

2. The other set without z-score scaling for both the training and test datasets.

I trained two models separately using these datasets and conducted predictions, but the results were completely opposite to my expectations!The dataset with scaling yielded evaluation results similar to what you observed, while the dataset without scaling produced results consistent with those in the original paper.
Thus, I now wonder if the preprocessing steps in the paper are rigorous and scaling is truly unnecessary—and could it be that you actually applied scaling?

**Comment by Jiram-Kin** [comment]
Hello there @siyuh, I wonder if there is any update regarding the code for reproducing the Fig. 2. I also encounter similarly low performance using the default training code (and I wasn't applying scaling but maybe it might be different in splitting/filtering as well?), it would be great if there are available code to reproduce this experiment.

Thank you for looking into this matter

**Comment by siyuh** [comment]
Hi, I have released the data preparation for fig2 as well as the checkpoints, please check https://github.com/siyuh/Squidiff_reproducibility for more details. I have train and run the model several times to confirm reproducing the results. If you still have issue to reproduce, please let me know. Thanks!
